### PR TITLE
Music WIP

### PR DESCRIPTION
**Pull request by bcjordan**
- Pulls MusicController up one level (from Studio to apps/)
- Play music a given delay after first Run button press
- Define a per-Craft-app-level song or set of songs to play through periodically

### CODE REVIEW

File: apps/src/MusicController.js [hunk 1]
```diff
@@ -89,9 +94,16 @@ module.exports = MusicController;
 function buildTrackData(trackDefinitions, assetUrl) {
   trackDefinitions = utils.valueOr(trackDefinitions, []);
   return trackDefinitions.map(function (trackDef) {
+
+    var assetUrls = [];
+    assetUrls.push(assetUrl(trackDef.name + '.mp3'));
+    if (trackDef.hasOgg) {
+      assetUrls.push(assetUrl(trackDef.name + '.ogg'));
+    }
```
Comment (islemaster) [inline apps/src/MusicController.js:134]: Seems good.  We probably should have gone back and put in ogg tracks for Star Wars too, but I'm not sure which browsers can play OGG and not MP3.

Please add the `hasOgg` to the MusicTrackDefinition typedef on line 27 of this file.

File: apps/src/craft/craft.js [hunk 2]
```diff
@@ -77,6 +78,16 @@ var interfaceImages = {
   ]
 };
 
+var musicMetadata = [
```
Comment (islemaster) [inline apps/src/craft/craft.js]: Should be a SCREAMING_SNAKE_CASE constant?

File: apps/src/craft/craft.js [hunk 3]
```diff
@@ -161,6 +176,42 @@ Craft.init = function (config) {
   Craft.level = config.level;
   Craft.skin = config.skin;
 
+  var levelTracks = [];
+  if (Craft.level.songs && musicMetadata) {
+    levelTracks = musicMetadata.filter(function(trackMetadata) {
+      return Craft.level.songs.indexOf(trackMetadata.name) !== -1;
+    });
```
Comment (islemaster) [inline apps/src/craft/craft.js:183]: Now that I see this in two places, I wonder if this `getMetadataForTrackNames` operation should be pulled into some kind of static helper on MusicController.

File: apps/src/craft/craft.js [hunk 4]
```diff
@@ -161,6 +176,42 @@ Craft.init = function (config) {
...
+  }
+
+  Craft.musicController = new MusicController(
+      studioApp.cdoSounds,
+      (filename) => config.skin.assetUrl(`music/${filename}`),
```
Comment (islemaster) [inline apps/src/craft/craft.js]: Bwuh? :confused: How is this ES6 working?

File: apps/src/craft/craft.js [hunk 5]
```diff
@@ -161,6 +176,42 @@ Craft.init = function (config) {
...
+      levelTracks,
+      levelTracks.length > 1 ? 7500 : null
+  );
+  if (studioApp.cdoSounds && !studioApp.cdoSounds.isAudioUnlocked()) {
+    // Would use addClickTouchEvent, but iOS9 does not let you unlock audio
+    // on touchstart, only on touchend.
+    var removeEvent = dom.addMouseUpTouchEvent(document, function () {
+      studioApp.cdoSounds.unlockAudio();
+      removeEvent();
+    });
```
Comment (islemaster) [inline apps/src/craft/craft.js:200]: This is another feature that would be nice to pull straight into soundEffects.js, now that I see it duplicated - but that may be risky at this point.

File: apps/src/craft/craft.js [hunk 6]
```diff
@@ -161,6 +176,42 @@ Craft.init = function (config) {
...
+  // Play music when the instructions are shown
+  var playOnce = function () {
+    if (studioApp.cdoSounds && studioApp.cdoSounds.isAudioUnlocked()) {
+      $(window).off('run_button_pressed', playOnce);
+
+      window.setTimeout(() => {
+        var hasSongInlevel = Craft.level.songs && Craft.level.songs.length > 1;
+        var songToPlayFirst = hasSongInlevel ? Craft.level.songs[0] : null;
+        Craft.musicController.play(songToPlayFirst);
+      }, Craft.level.songDelay || 1500);
```
Comment (islemaster) [inline apps/src/craft/craft.js]: :+1: Nicely done - so you enforce the first song, then let it pick randomly after that, right?

**Comment by islemaster** [comment]
Here's a scenario you might have missed:
1. Start a level with a video.
2. Close the video - music should start.
3. Re-show the video - music will fade out and stop.
4. In Star Wars the music never comes back in this scenario, and we're fine with that since we don't loop it anyway.  In your case, I think another music track might start playing during the video after a delay.  You may need to extend MusicController to have some sort of 'paused until video hidden' state.

**Comment by bcjordan** [comment]
Good thinking. I'm seeing it work OK on Chrome at least, `whenMusicStopped_` doesn't get hit when faded out by the video since the video fade out calls `this.stop()`. But since it's possible that call might cause another browser to fire off the stopped event, it does sound worth hardening with more stop/resume repeating behavior

File: apps/src/craft/craft.js (hunk 6 above)
Comment (bcjordan) [inline apps/src/craft/craft.js]: That's right. Next up will be "don't repeat the same song that just played", but may not end up with time to get that in today

**Comment by islemaster** [comment]
Oh, I thought `whenMusicStopped_` would get fired in that scenario (I think it's supposed to happen no matter what stops the music?).  In that case, do you need to add logic to resume the music when the video is closed a second time?

File: apps/src/MusicController.js (hunk 1 above)
Comment (bcjordan) [inline apps/src/MusicController.js:134]: Done!

File: apps/src/craft/craft.js (hunk 2 above)
Comment (bcjordan) [inline apps/src/craft/craft.js]: Done!

File: apps/src/craft/craft.js (hunk 4 above)
Comment (bcjordan) [inline apps/src/craft/craft.js]: :innocent: I'll replace with normal callback until we discuss ES6 usage / style next week

Comment (islemaster) [inline apps/src/craft/craft.js]: I'm totally into it, I just didn't know our pipeline supported it yet.

**Comment by islemaster** [comment]
LGTM

File: apps/src/MusicController.js [hunk 7]
```diff
@@ -71,9 +76,41 @@ var MusicController = function (audioPlayer, assetUrl, trackDefinitions) {
   /** @private {string} Name of track to play on load */
   this.playOnLoad_ = null;
 
+
+  /** @private {number} */
+  this.loopRandomWithDelay_ = loopRandomWithDelay;
+
+  /**
+   * @private {boolean} whether we stopped playing music due to video being
+   *          shown
```
Comment (islemaster) [inline apps/src/MusicController.js:85]: If you want to capture the double-meaning of this variable, could at least do it in this comment for now.

Comment (islemaster) [inline apps/src/MusicController.js:85]: IsVideoShowingAndDidItInterruptPlayback

File: apps/src/MusicController.js [hunk 8]
```diff
@@ -219,5 +271,6 @@ MusicController.prototype.getTrackByName_ = function (name) {
  * @private
  */
 MusicController.prototype.getRandomTrack_ = function () {
-  return this.trackList_[Math.floor(Math.random(this.trackList_.length))];
+  var trackIndex = Math.floor(Math.random() * this.trackList_.length);
+  return this.trackList_[trackIndex];
```
Comment (islemaster) [inline apps/src/MusicController.js:275]: :+1: Thanks for fixing this bug!

**Comment by islemaster** [comment]
LGTMAgain

**Pull request merged**